tests/installer/default.nix

```diff
@@ -0,0 +1,101 @@
+# Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+#
+# This is a main test for ghaf. If you want to debug it interactively
+# you can follow this guide https://blog.thalheim.io/2023/01/08/how-to-execute-nixos-tests-interactively-for-debugging/.
+# In other words, call breakpoint() at the point of testScript
+# that you interested in to debug. After that you can build test driver using
+# `nix build .#checks.x86_64-linux.installer.driver` and then run driver
+# `./result/bin/nixos-test-driver`. Wait until the moment logs stop appear,
+# after that you can execute `interact` and finally `new_machine.shell_interact()`.
+# This will allow you to interact with installed ghaf.
+{ pkgs, self }:
+let
+  testConfig = "lenovo-x1-carbon-gen11-debug";
+  system = "x86_64-linux";
+  expectedHostname = "ghaf-host";
+
+  cfg = self.nixosConfigurations.${testConfig};
+
+  testingConfig = cfg.extendModules {
+    modules = [
+      (cfg._module.specialArgs.modulesPath + "/testing/test-instrumentation.nix")
+      (cfg._module.specialArgs.modulesPath + "/profiles/qemu-guest.nix")
+      (_: {
+        testing.initrdBackdoor = true;
+        services.openssh.enable = true;
+        # https://github.com/nix-community/disko/blob/e55f9a8678adc02024a4877c2a403e3f6daf24fe/lib/interactive-vm.nix#L63
+        boot.zfs.devNodes = "/dev/disk/by-uuid"; # needed because /dev/disk/by-id is empty in qemu-vms
+        boot.zfs.forceImportAll = true;
+      })
+    ];
+  };
+
+  # FIXME: Only one attribute supported. What about ISO?
+  imagePath = testingConfig.config.system.build.diskoImages + "/disk1.raw.zst";
+  targetPath = "/dev/vdb";
+  installScript = pkgs.callPackage ../../packages/installer { };
+  installerInput = pkgs.lib.strings.escapeNixString "${targetPath}\ny\ny\n";
+in
+pkgs.nixosTest {
+  name = "installer-test";
+  nodes.machine = {
+    virtualisation.emptyDiskImages = [ (1024 * 256) ];
```

Comment:
Oh, that's quite huge :)
I hope that doesn't break CI.
Maybe it could be made a bit smaller for the test only?
You can introduce also a nixos option that only gets set here.

Reply:
It's fine, this is just upper bound for the size of a qcow2 file as it allocates memory only when needs it. After installation `empty0.qcow2` size is only 16GB.

Reply:
I assume this is a sparse file than.